app/models/cost_query/cache.rb

```diff
@@ -0,0 +1,75 @@
+#-- copyright
+# OpenProject Reporting Plugin
+#
+# Copyright (C) 2010 - 2014 the OpenProject Foundation (OPF)
+#
+# This program is free software; you can redistribute it and/or
+# modify it under the terms of the GNU General Public License
+# version 3.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program; if not, write to the Free Software
+# Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+#++
+
+module CostQuery::Cache
+  class << self
+
+    def check
+      reset! if reset_required?
+    end
+
+    def reset!
+      CostQuery::Filter.reset!
+      CostQuery::Filter::CustomFieldEntries.reset!
+      CostQuery::GroupBy.reset!
+      CostQuery::GroupBy::CustomFieldEntries.reset!
+
+      update_reset_on
+    end
+
+    protected
+
+    attr_accessor :custom_fields_updated_on,
+                  :custom_fields_id_sum
+
+    def invalid?
+      updated_on = fetch_custom_field_updated_at
+      id_sum = fetch_custom_fields_changed
+
+      custom_fields_updated_on != updated_on ||
+        custom_fields_id_sum != id_sum
+    end
+
+    def update_reset_on
+      return if caching_disabled?
+
+      self.custom_fields_updated_on = fetch_custom_field_updated_at
+      self.custom_fields_id_sum = fetch_custom_fields_changed
+    end
+
+    def fetch_custom_field_updated_at
+      WorkPackageCustomField.maximum(:updated_at)
+    end
+
+    def fetch_custom_fields_changed
+      WorkPackageCustomField.sum(:id) + WorkPackageCustomField.count
+    end
+
+    def caching_disabled?
+      !OpenProject::Configuration.cost_reporting_cache_filter_classes
+    end
+
+    def reset_required?
+      caching_disabled? || invalid?
```

Comment:
What the code does is to dynamically generate classes based on the CustomFields that are required. Those classes are then referenced by a couple of other classes via their `all` methods. That is all the "caching" there is. So the only decision we take here is whether we want to get rid of the dynamically created classes (`reset!`) or not.

We reference those `all` methods at a couple of places throughout a request. So it is a sure bet that creating the classes multiple times for one request is bad performancewise. 

The question then boils down to do whether we expect it to be good if we keep those classes and references over multiple requests. And this to me boils down to the question of whether we expect the same set of CustomFields in the next request. If the set of CustomFields change, we have to get rid of the created classes so that we display the correct filters. Please note, that checking whether the classes are still valid does have a performance impact as well (multiple database hits). So checking, noticing that the set changed and then creating the classes is more expensive than only creating the classes.

For a single database setting where only one instance is served, it is more likely to have the same set of CustomFields in the current request, compared to the last, so the overhead of checking the class validity is justified. However, if we serve multiple instances from one process, it becomes highly likely that the next request is not directed at the same instance/database/set of custom fields. One can therefore disable the check and throw the classes away without it.

Granted, the probability of not getting the same user in the next request is only an opinion I have.